### ISSUE TITLE
Invalid strides for dim with 1 element

### ISSUE THREAD

**Issue by ogrisel** [numpy/numpy#7228]
```python
>>> import numpy as np
>>> np.ones(1).strides
(9223372036854775807,)
```

The strides are correct in numpy 1.10.4.

**Comment by ogrisel** [comment]
Hum this is apparently intentional to have this in the dev branch: #6342.

However why would inspecting the strides be illegal? Here is the crash we get in scikit-learn:

https://github.com/scikit-learn/scikit-learn/issues/6334

More precisely here our code:

https://github.com/scikit-learn/scikit-learn/blob/0.17/sklearn/utils/seq_dataset.pyx#L196

Is this an abuse of the numpy API?

**Comment by pv** [comment]
AFAIK:
- Inspecting the strides is OK.
- For 1-element arrays, any strides is a valid value, since `0*x==0` for integer `x`.
- A relation `element_size = stride[0] / size` is not guaranteed.
- There was a motivation for this change in that it allowed more efficient operation in some cases --- I don't remember the details though.
- The data type for strides is `intp`. `int` may be too small (2**32 is only ~4GB). If the sklearn Cython code uses `int` variables for strides, this can be understood as a bug.

**Comment by charris** [comment]
It is part of the relaxed strides transition and, IIRC, intended to smoke out misuse of strides to determine contiguity. @seberg Comment?

**Comment by njsmith** [comment]
Looking at `strides` is totally legal; the `OverflowError` is unintentional, and actually I don't understand where it's even coming from looking at that code -- some mysterious cython thing I guess.

I also don't understand what that code is even trying to do. Is the goal of that line to try and compute `X.shape[1]` or something? Anyway I guess it'd be fine if cython hadn't blown up for no reason -- the only thing you do with `stride` is multiply by an index down below, and notice that for this position the first dimension has shape 1, so the only valid index is 0.

The only guarantee that numpy makes about strides is that if `index` is a valid index (`all(index <= shape)`), then `index @ strides` will give you a valid offset. Unfortunately for contiguous arrays it used to make somewhat stronger guarantees that didn't make much sense, and created weird paradoxes like saying that you could have two row vectors that had identical memory layout but one was C contiguous but not F contiguous, and the other was F contiguous but not C contiguous. The special value is to try to smoke out code that depends on these stronger assumptions, and does things like `itemsize = a.strides[-1]` is a valid way to compute itemsize (wrong, don't do this even if `a` has the C contiguous flag set).

**Comment by seberg** [comment]
Yeah, the special code is more to debug, less for production (I always
somewhat thought we could at some point have a switch to do the weird
stuff possibly during dev but not release versions or so.

Maybe there is some signedness issue, since it is INT_MIN and not
INT_MAX IIRC. Nothing would be wrong about using a slightly less weird
bad value.

**Comment by ogrisel** [comment]
> I always somewhat thought we could at some point have a switch to do the weird stuff possibly during dev but not release versions or so.

I tested scikit-learn against the master branch of numpy because I would like to continuously build sklearn master against numpy and scipy master to detect potential regressions earlier in our build process.

So it would be great to have this weird stuff for numpy dev not enabled at build time but rather via a runtime switch (e.g. a dedicated environment variable). WDYT?

**Comment by njsmith** [comment]
This weird thing is exactly intended for shaking out bugs in downstream projects like yourself -- when we first turned this on it e.g. triggered several crashes in scipy that were real (latent) bugs.

Before changing anything in numpy I think we should first understand what's really going on with the original code, since the `OverflowError` sounds like it probably indicates a real bug for large matrices, and the original code is suspicious-looking in other ways...

**Comment by ogrisel** [comment]
Indeed, as @pv commented earlier the cause of the issue is the use of the `int` type instead of `np.npy_intp` in our cython code. I just checked, and fixing the type make the overflowerror go away.

Thank you for your help.
